### ISSUE TITLE
Extend option-map-unwrap-or-else to Result

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust-clippy#1590]
`opt.map(f).unwrap_or_else(g)` is linted to suggest `opt.map_or_else(g, f)`, but `res.map(f).unwrap_or_else(g)` is not linted even though it could be `res.ok().map_or_else(|_| g(), f)`

Somewhat ugly suggestion, so should be allow-by-default, maybe we should poke rustc to add `Result::map_or_else`

**Comment by Arnavion** [comment]
This was implemented in #2189, so it could be closed, though I'll note that both the OP suggestion and the implementation of the lint are wrong in the *general* case. `res.map(f).unwrap_or_else(g)` cannot be replaced with `res.ok().map_or_else(g, f)` if `g` uses the `Err` value. The lint should be tightened to only fire if `g` is `|_|`

**Comment by mathstuf** [comment]
Why is it wrong? `f` doesn't get called if `res` is `Err`, so it goes right to `g`. If it is `Ok`, `g` isn't getting called at all. Maybe you confused it with `and_then` where `g` could get `res` or `f`'s `Err` depending on the behavior of `f`?

`map` followed by `unwrap_or_else`:

```rust
let intermediate = match res {
  Ok(v) => Ok(f(v)),
  Err(err) => Err(err),
};
match intermediate {
  Ok(v) => v,
  Err(err) => g(err),
}
```

`unwrap_or_else`:

```rust
match res {
  Ok(v) => f(v),
  Err(err) => g(err),
}
```

These look isomorphic to me.

@oli-obk Good to close?

**Comment by Arnavion** [comment]
@mathstuf If you were responding to my comment, you should read more closely what I wrote and the lint is suggesting you to use. Notice that it's not suggesting `Result::map_or_else`

**Comment by mathstuf** [comment]
Ah, missed the `.ok()` that's in there. If we just get rid of that suggestion, this is resolved then, right?
